kyuubi-ha/src/test/scala/org/apache/kyuubi/ha/client/v2/EngineServiceDiscoveryClientSuite.scala

```diff
@@ -0,0 +1,158 @@
+import org.apache.kyuubi.ha.v2.engine.strategies.SessionNumStrategy
+import org.apache.kyuubi.zookeeper.{EmbeddedZookeeper, ZookeeperConf}
+
+class EngineServiceDiscoveryClientSuite extends KyuubiFunSuite {
+
+  protected val conf: KyuubiConf = KyuubiConf()
+
+  private var zkServer: EmbeddedZookeeper = _
+
+  override def beforeAll(): Unit = {
+    zkServer = new EmbeddedZookeeper()
+    val zkData = Utils.createTempDir()
+    conf.set(ZookeeperConf.ZK_DATA_DIR, zkData.toString)
+    zkServer.initialize(conf)
+    zkServer.start()
+    conf.set(HighAvailabilityConf.HA_ZK_NAMESPACE.key, s"localhost:${ZookeeperConf.ZK_CLIENT_PORT}")
+  }
+
+  override def afterAll(): Unit = {
+    if (zkServer != null) {
+      zkServer.stop()
+      zkServer = null
+    }
+    super.afterAll()
+  }
+
+  test("engine discovery client test") {
+    val namespace = "/kyuubi_USER/test"
+    ServiceDiscovery.withZkClient(conf) { zkClient =>
+      val engine1 = EngineInstance(namespace, null, "127.0.0.1", 1111,
+        Some("1.2.0"), 2, Seq("SPARK_3", "p0"))
+      val path1 = createZkServiceNode(conf, zkClient, engine1).getActualPath
+
+      val engine2 = EngineInstance(namespace, null, "127.0.0.1", 2222,
+        Some("1.3.0"), 1, Seq("SPARK_3", "p1"))
+      val path2 = createZkServiceNode(conf, zkClient, engine2).getActualPath
+
+      val engine3 = EngineInstance(namespace, null, "127.0.0.1", 3333,
+        Some("1.3.0"), 0, Seq("SPARK_3", "p2"))
+      val path3 = createZkServiceNode(conf, zkClient, engine3).getActualPath
+
+      // session number base
+      conf.set(ServiceDiscoveryConf.ENGINE_PROVIDER_STRATEGY, SessionNumStrategy.NAME)
+      val instance1 = EngineServiceDiscoveryClient.get(namespace, zkClient, conf)
+        .getInstance().get
+      assert(instance1.instance.equals(engine3.instance))
+
+      // version filter
+      conf.set(ServiceDiscoveryConf.ENGINE_PROVIDER_VERSION, "1.2.0")
+      val instance2 = EngineServiceDiscoveryClient.get(namespace, zkClient, conf).getInstance().get
+      assert(instance2.instance.equals(engine1.instance))
+
+      // tags filter
+      conf.unset(ServiceDiscoveryConf.ENGINE_PROVIDER_VERSION)
+      conf.set(ServiceDiscoveryConf.ENGINE_PROVIDER_TAGS, "SPARK_3,p1")
+      val instance3 = EngineServiceDiscoveryClient.get(namespace, zkClient, conf).getInstance().get
+      assert(instance3.instance.equals(engine2.instance))
+    }
+  }
+
+  def createZkServiceNode(
+      conf: KyuubiConf,
+      zkClient: CuratorFramework,
+      engine: EngineInstance): PersistentNode = {
+    val ns = ZKPaths.makePath(null, engine.namespace)
+    try {
+      zkClient
+        .create()
+        .creatingParentsIfNeeded()
+        .withMode(PERSISTENT)
+        .forPath(ns)
+    } catch {
+      case _: NodeExistsException => // do nothing
+      case e: KeeperException =>
+        throw new KyuubiException(s"Failed to create namespace '$ns'", e)
+    }
+    val pathPrefix = ZKPaths.makePath(
+      engine.namespace,
```

Comment:
Currently, the EngineSpace(namespace) and serviceNode(engine instance) are 1 vs 1

Reply:
Then, the ` EngineSpace` could achieve the isolation purpose for a tenant.

```scala
  /**
   * The EngineSpace used to expose itself to the KyuubiServers in `serverSpace`
   *
   * For `CONNECTION` share level:
   *   /$serverSpace_CONNECTION/$user/$sessionId
   * For `USER` share level:
   *   /$serverSpace_USER/$user[/$subDomain]
   *
   */
```

Reply:
The issue I mentioned in #786 is that for now we can not ensure the `serviceNode` here is alive or not

Reply:
> Currently, the EngineSpace(namespace) and serviceNode(engine instance) are 1 vs 1

It doesn't seem to be one-to-one, engin namespace like: `/kyuubi_USER/test`, instance path like `/kyuubi_USER/test/serviceUri=127.0.0.1:1234;version=1.3.0;sequence=`

Reply:
> The issue I mentioned in #786 is that for now we can not ensure the serviceNode here is alive or not

Sorry, I haven't considered it. This PR is mainly to choose the more appropriate one for all engines under the same shared user.

Reply:
In this test, yes, it's not one-to-one. but in real-world usage, we have a distributed lock for engine creations at the server-side.

Reply:
I added a `CoreSizeStrategy` to create more instances.  When the number of engines does not reach `CoreSize`, do not return to the existing engine, but to create a new engine.

Reply:
I get your point. How about when the client-side uses `DBCP`? the connections will point to different engines, will this result in surprise?

Reply:
> I get your point. How about when the client-side uses DBCP? the connections will point to different engines, will this result in surprise?

Yes, this may be problematic, but we can use a unique tag to adapt it.

Reply:
Does it allow only one engine for a unique tag?

Reply:
then the tag is a bit redundant with `[/$subDomain]`, we need to remove one of them

Reply:
Sorry, it does not seem to be the case, it may also need to limit core size.

Reply:
Can we minimize the PR scope and do these separately?

Maybe we can start by adding a pool size for engines, and randomly pick it. then we apply picking rules to it separately in a series of PRs

Reply:
another thing is that we don't want the engine discovery to be bind to zookeeper too much in the near future, we are seeking a more light way to do this

Reply:
OK, thank you. I will minimize it and implement it gradually.